docs/get-started/support-new-model.md

```diff
@@ -0,0 +1,548 @@
+                    relative_position_scores_key = torch.einsum("bhrd,lrd->bhlr", key_layer, positional_embedding)
+                    attention_scores = attention_scores + relative_position_scores_query + relative_position_scores_key
+
+            attention_scores = attention_scores / math.sqrt(self.attention_head_size) # (2)
+            if attention_mask is not None:
+                # Apply the attention mask is (precomputed for all layers in BertModel forward() function)
+                attention_scores = attention_scores + attention_mask
+
+            # Normalize the attention scores to probabilities. # (3)
+            attention_probs = nn.functional.softmax(attention_scores, dim=-1)
+
+            # This is actually dropping out entire tokens to attend to, which might
+            # seem a bit unusual, but is taken from the original Transformer paper.
+            attention_probs = self.dropout(attention_probs)
+
+            # Mask heads if we want to
+            if head_mask is not None:
+                attention_probs = attention_probs * head_mask
+
+            context_layer = torch.matmul(attention_probs, value_layer) # (4)
+
+            context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
+            new_context_layer_shape = context_layer.size()[:-2] + (self.all_head_size,)
+            context_layer = context_layer.view(new_context_layer_shape)
+
+            outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+
+            if self.is_decoder:
+                outputs = outputs + (past_key_value,)
+            return outputs
+    ```
+
+    1.  $QK^T$
+    2.  $\frac{QK^T}{\sqrt{d_k}}$
+    3.  $\operatorname{softmax}(\frac{QK^T}{\sqrt{d_k}})$
+    4.  $\operatorname{softmax}(\frac{QK^T}{\sqrt{d_k}})V$
+
+We see that the Hugging Face implementation is close to the definition in the paper, we can actually want to find the attention pattern in this model.
+
+To begin, we'll write a short script running the model with a dummy input:
+
+``` { .py }
+import torch
+from transformers import AutoModel
+
+from kernl.model_optimization import optimize_model
+
+
+model = AutoModel.from_pretrained(pretrained_model_name_or_path="bert-base-uncased").eval().cuda()
+
+optimize_model(model)
+shape = (1, 128)
+
+with torch.inference_mode(), torch.cuda.amp.autocast(enabled=True, dtype=torch.float16, cache_enabled=True):
+    inputs = {
+        "input_ids": torch.randint(2, 10000, shape, device="cuda", dtype=torch.long),
+        "attention_mask": torch.ones(shape, device="cuda", dtype=torch.long),
+    }
+    output = model(**inputs)
+```
+
+If we run `#!python graph_report(gm)` and `#!python gm.code` in the `#!python dynamo_backend_ofi(gm)` function, we can print the FX graph and the python code from the model computation. For our example, we'll keep the `normalize_operators` and the `remove_dropout` functions as it simplifies a bit the model's graph.
```

Comment:
Are you suggesting the user should modify our code ? Maybe the tutorial should be written in such a way we can do everything without modifying our code, but I'm not sure if it's possible with current codebase

Reply:
It's not really possible with the actual codebase.  I was thinking of two ways.

The first one is to write a jupyter notebook where we only overload the optimization function and play with it. But a jupyter notebook is not really easy for a first time read.

The second one was to adapt our codebase easily add / remove optimization (like a pluggin system), and one of the "plugin" only prints the graph. But it demands a lot of work on the code.